metal/resource_metal_device.go

```diff
@@ -348,8 +350,66 @@ func resourceMetalDevice() *schema.Resource {
 				Default:     false,
 				ForceNew:    false,
 			},
+			"reinstall": {
+				Type:     schema.TypeList,
+				MaxItems: 1,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"enabled": {
+							Type:        schema.TypeBool,
+							Description: "Whether the device should be reinstalled instead of destroyed",
+							Optional:    true,
+							Default:     false,
+						},
+						"deprovision_fast": {
+							Type:        schema.TypeBool,
+							Description: "Whether the OS disk should be filled with `00h` bytes before reinstall",
+							Optional:    true,
+							Default:     false,
+						},
+						"preserve_data": {
+							Type:        schema.TypeBool,
+							Description: "Whether the non-OS disks should be kept or wiped during reinstall",
+							Optional:    true,
+							Default:     false,
+						},
+					},
+				},
+			},
 		},
+		CustomizeDiff: customdiff.Sequence(
+			customdiff.ForceNewIf("custom_data", shouldReinstall),
+			customdiff.ForceNewIf("operating_system", shouldReinstall),
+			customdiff.ForceNewIf("user_data", shouldReinstall),
+		),
+	}
+}
+
+func shouldReinstall(_ context.Context, d *schema.ResourceDiff, meta interface{}) bool {
+	reinstall, ok := d.GetOk("reinstall")
+
+	// Prior behaviour was to always destroy and create,
+	// so in the event we can't get the reinstall config; let's
+	// continue to do so.
+	if !ok {
+		return true
+	}
+
+	// We didn't get a reinstall configuration
+	reinstall_list, ok := reinstall.([]interface{})
+	if !ok {
+		return true
+	}
+
+	reinstall_config, ok := reinstall_list[0].(map[string]interface{})
```

Comment:
I'm sorry you had to experience this 😅 
Too bad the SDK doesn't provide `d.GetFirstMapStringOk()` (heh).